barrels/LogMonkey/LogTests.mc

```diff
@@ -0,0 +1,246 @@
+using Toybox.Test as Test;
+using Toybox.Lang as Lang;
+using LogMonkey as Log;
+
+module LogMonkey {
+
+    // Wrap the unit tests in a module annotated with :test so it will get compiled
+    // out for non unit test builds.
+    (:test)
+    module Test {
+
+        //! This class mocks the Toybox.System.println() interface but instead
+        //! of printing a value saves it to a string value which can be retrieved
+        //! and examed to verify log results are correct.
+        class MockSys {
+
+            private var mStreamContents;
+
+            //! A mock of Toybox.System.println() function
+            //! @param [Toybox::Lang::Object] message The message to print/log
+            public function println(message) {
+                if (mStreamContents == null) {
+                    mStreamContents = message + "\n";
+                } else {
+                    mStreamContents = mStreamContents + message + "\n";
+                }
+            }
+
+            //! Gets the contents of the fake print stream.
+            //! @return [Toybox::Lang::String] The contents that have been "printed" so far
+            public function getStreamContents() {
+                return mStreamContents;
+            }
+
+        }
+
+        // This class is used within the unit tests module to create an Exception
+        // object with a custom message.
+        class TestException extends Lang.Exception {
+            function initialize() {
+                Exception.initialize();
+                mMessage = "This is a test Exception";
+            }
+        }
+
+        //! Tests determing the type of a variable
+        (:test)
+        function testDetermingVariableType(logger) {
+            var mockSys;
+            var log;
+            var logOutput;
+            var level = "D";
+            var tag = "TAG";
+            var varName = "varToTest";
+            var varToTest;
+
+            // Test null
+            varToTest = null;
+            mockSys = new MockSys();
+            log = new $.LogMonkey.Logger(level, mockSys);
+            log.logVariable(tag, varName, varToTest);
+            logOutput = mockSys.getStreamContents();
+            logger.debug("log output: " + logOutput);
+            Test.assert(logOutput.find("(null)") != null);
+
+            // Test Lang.Array
+            varToTest = [1, 2, 3];
+            mockSys = new MockSys();
+            log = new $.LogMonkey.Logger(level, mockSys);
+            log.logVariable(tag, varName, varToTest);
+            logOutput = mockSys.getStreamContents();
+            logger.debug("log output: " + logOutput);
+            Test.assert(logOutput.find("(Array)") != null);
```

Comment:
Should the testing be more thorough and verify the actual value output matches what we expect? If so, perhaps there should be a test function for verifying each variable type we support?

Reply:
I'd be open to looking for a more exact value if you have suggestions on how to do it. I really just wanted a test to ensure that when given a variable type to getVariableType() we'd get the expected value in parentheses in the output message. The problem I was running into is that the log format includes a timestamp so we can do an exact match and there's not a great way to do a regex type of check. We could have a function that breaks the message down and looks for a specific part to match exactly but I don't know if that will provide better value that just looking for the specific type piece.

Reply:
Well, MonkeyC doesn't have regular expression support, but I was thinking you could do something like this..

```
(:test)
module Test {

    // test helper function 
    function do_testLogVariable(level, tag, name, variable, match) {

        var mockSys = new MockSys();

        var log = new $.LogMonkey.Logger(level, mockSys);
        log.logVariable(tag, name, variable);
        log = null;

        var logOutput = mockSys.getStreamContents();

        // look for the caller-provided substring for an exact match
        Test.assert(logOutput.find(match) != null);
    }

    (:test)
    function testLogVariable_Number(logger) {
        do_testLogVariable("D", "x", "aNumber", -1, " {D} x: aNumber=(Number) -1");
        do_testLogVariable("W", "y", "bNumber", 0, " {W} y: bNumber=(Number) 0");
        do_testLogVariable("E", "z", "cNumber", 1, " {E} y: cNumber=(Number) 1");
        // test more values if we feel it is necessary (INT_MIN, INT_MIN + 1, INT_MAX - 1, INT_MAX)
    }

    (:test)
    function testLogVariable_Array(logger) {
        do_testLogVariable("E", "a", "xArray",           [], " {E} a: xArray=(Array) []");
        do_testLogVariable("W", "b", "yArray", [ 1, "abc" ], " {W} b: yArray=(Array) [1, abc]");
        // test more values if we feel it is necessary
    }

    (:test)
    function testLogVariable_Prefix(logger) {
        var mockSys = new MockSys();

        var log = new $.LogMonkey.Logger("I", mockSys);
        log.logVariable("i", "iI", logger);
        log = null;

        var logOutput = mockSys.getStreamContents();

        // this assert may have to change when the version changes
        // verify the version happens at the front
        Test.assert(logOutput.find("(lmf1)" == 0));
    }
}
```

If you wanted to properly handle the date/time you'd have to give the logger something that it could use to get the date/time and then mock it in the tests just as you've done with ``Sys.println()``. Given what you've got right now, the pattern would be to take two additional parameters (``Gregorian`` and ``Time``).

Reply:
I split up the test types unit test into the individual variable types using the helper function you suggested. I'm going to leave the mocked time values out as I don't want to have the unit test dictate the constructor interface for the Logger class.